File: python/server/response_sender.py
from model.response import Response

import json as json_lib
from http.server import BaseHTTPRequestHandler
from http import HTTPStatus


def sendWinnerResponse(r: BaseHTTPRequestHandler, response: Response):
    """リクエスト元にオークション情報をJSON形式にして返す

    Args:
        r (BaseHTTPRequestHandler): http.serverのリクエスト情報
        response (Response): オークションに勝った広告情報などが保存されているモデル
    """
    json = {"auction_id": response.auction_id, "ads": response.ads, "price": response.price}
    data = json_lib.dumps(json, separators=(",", ":"))
    r.send_response(HTTPStatus.OK)
    r.end_headers()
    r.wfile.write(data.encode())


def sendNobidResponse(r: BaseHTTPRequestHandler):
    """該当する広告がない場合に `204 No Contents` を返す

    Args:
        r (BaseHTTPRequestHandler): http.serverのリクエスト情報
    """
    r.send_response(HTTPStatus.NO_CONTENT)
    r.end_headers()

File: cpp/src/server/response_sender.cpp
#pragma once

#include <event2/buffer.h>
#include <event2/http.h>
#include <rapidjson/stringbuffer.h>
#include <rapidjson/writer.h>

#include <memory>

#include "../model/response.cpp"

namespace geniee::server {
    /// @brief リクエスト元にオークション情報をJSON形式にして返す
    /// @param r libeventのリクエスト情報
    /// @param response オークションに勝った広告情報などが保存されているモデル
    void sendWinnerResponse(evhttp_request *r, const geniee::model::Response response) {
        rapidjson::StringBuffer buffer;
        rapidjson::Writer<decltype(buffer)> writer(buffer);
        writer.StartObject();
        writer.Key("auction_id");
        writer.String(response.getAuctionID());

        writer.Key("ads");
        writer.String(response.getAds());

        writer.Key("price");
        writer.Double(response.getPrice());

        writer.EndObject();

        std::unique_ptr<evbuffer, decltype(&evbuffer_free)> buf(evbuffer_new(), &evbuffer_free);
        if (!buf) {
            evhttp_send_error(r, HTTP_SERVUNAVAIL, "Internal Server Error");
            return;
        }

        evbuffer_add(buf.get(), buffer.GetString(), buffer.GetSize());
        evhttp_send_reply(r, HTTP_OK, "OK", buf.get());
    }

    /// @brief 該当する広告がない場合に 204 No contents を返す
    /// @param r libeventのリクエスト情報
    void sendNobidResponse(evhttp_request *r) {
        evhttp_send_reply(r, HTTP_NOCONTENT, nullptr, nullptr);
    }

}  // namespace geniee::server
